File: medium/547. Number of Provinces/547. Number of Provinces.ts
function findCircleNum(provinces: number[][]): number {
    const visited = new Set()
    const n = provinces.length
    let res = 0

    const dfs = (city: number) => {
        if (visited.has(city)) return
        visited.add(city)
        for (let conn = 0; conn < n; conn) {
            if (provinces[city][conn] == 1)
                dfs(conn)
        }
    }

    for (let city = 0; city < n; city++) {
        if (visited.has(city)) continue
        dfs(city)
        res += 1
    }

    return res
};

File: medium/547. Number of Provinces/547. Number of Provinces.py
from typing import List


class Solution:
    def findCircleNum(self, connections: List[List[int]]) -> int:
        n = len(connections)
        res = 0

        visited = set()

        def dfs(city: int):
            if city > n or city in visited:
                return

            visited.add(city)
            for i, conn in enumerate(connections[city]):
                if conn:
                    dfs(i)

        for city in range(n):
            if city in visited:
                continue
            dfs(city)
            res += 1

        return res
